weave/trace_server/table_query_builder.py

```diff
@@ -23,30 +23,33 @@ def make_natural_sort_table_query(
     """
     project_id_name = pb.add_param(project_id)
     digest_name = pb.add_param(digest)
-    sql_safe_dir = "ASC" if natural_direction == "ASC" else "DESC"
 
-    sql_safe_limit = (
-        f"LIMIT {{{pb.add_param(limit)}: Int64}}" if limit is not None else ""
-    )
-    sql_safe_offset = (
-        f"OFFSET {{{pb.add_param(offset)}: Int64}}" if offset is not None else ""
-    )
+    row_digests_selection = "row_digests"
+    if natural_direction.lower() == "desc":
+        row_digests_selection = f"reverse({row_digests_selection})"
+    if limit is not None and offset is None:
+        offset = 0
+    if offset is not None:
+        if limit is None:
+            row_digests_selection = f"arraySlice({row_digests_selection}, 1 + {{{pb.add_param(offset)}: Int64}})"
+        else:
+            row_digests_selection = f"arraySlice({row_digests_selection}, 1 + {{{pb.add_param(offset)}: Int64}}, {{{pb.add_param(limit)}: Int64}})"
 
     query = f"""
     SELECT DISTINCT tr.digest, tr.val_dump, t.row_order
     FROM table_rows tr
-    RIGHT JOIN (
+    INNER JOIN (
```

Comment:
This is one key to the solution (INNER vs RIGHT)

Reply:
`INNER` is intersection, `RIGHT` is all of the second table + intersection. We can get away with using just the intersection? Intuitively this will now be fewer rows.

Reply:
Yes, the edge case here is if we accidentally deleted a row from the table_rows that was referenced in the table itself. This would be a bad error. However, most of the perf in this PR comes from this `INNER` clause...